src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs

```diff
@@ -355,37 +355,51 @@ internal static void UnshimResponseStream(ResponseCachingContext context)
         internal static bool ContentIsNotModified(ResponseCachingContext context)
         {
             var cachedResponseHeaders = context.CachedResponseHeaders;
-            var ifNoneMatchHeader = context.TypedRequestHeaders.IfNoneMatch;
+            var ifNoneMatchHeader = context.HttpContext.Request.Headers[HeaderNames.IfNoneMatch];
 
-            if (ifNoneMatchHeader != null)
+            if (!StringValues.IsNullOrEmpty(ifNoneMatchHeader))
             {
-                if (ifNoneMatchHeader.Count == 1 && ifNoneMatchHeader[0].Equals(EntityTagHeaderValue.Any))
+                if (ifNoneMatchHeader.Count == 1 && ifNoneMatchHeader[0].Equals(EntityTagHeaderValue.Any.Tag))
                 {
                     context.Logger.LogNotModifiedIfNoneMatchStar();
                     return true;
                 }
 
-                if (cachedResponseHeaders.ETag != null)
+                if (!StringValues.IsNullOrEmpty(cachedResponseHeaders[HeaderNames.ETag]))
                 {
-                    foreach (var tag in ifNoneMatchHeader)
+                    EntityTagHeaderValue eTag;
+                    if (EntityTagHeaderValue.TryParse(cachedResponseHeaders[HeaderNames.ETag], out eTag))
                     {
-                        if (cachedResponseHeaders.ETag.Compare(tag, useStrongComparison: false))
+                        foreach (var tag in ifNoneMatchHeader)
```

Comment:
I don't think you can use a `foreach` here since it's not guaranteed that each string contains only one ETag due how StringValues from HTTP headers are stored. Let's take a look at how it is parsed in Abstractions.